src/boxerror_replacement.rs

```diff
@@ -1,11 +1,40 @@
-//! Experimental replacement for `Box<dyn Error>` that implements `Error`
-use std::error::Error;
+//! Experimental replacement for `Box<dyn Error>` that implements `Error`.
+//!
+//! This module introduces the `DynError` type, which owns an inner `Box<dyn Error>`. Most
+//! importantly, `DynError` _does_ implement the `Error` trait, unlike `Box<dyn Error>`. As a
+//! result, `DynError` is more conveniently compatible with the rest of the error handling
+//! ecosystem, and behaves as any other error type does. 
+//!
+//! The short answer as to why `Box<dyn Error>` doesn't implement `Error` is because there 
+//! exists a blanket implementation of the `Error` trait for `Box<T>`, more specifically: 
+//! `impl<T: Error + Sized> Error for Box<T>`. Crucially, any type `T` must be sized. However, 
+//! when `T` is a `dyn Error` trait object, it is _not_ sized. This constraint could be loosened by
+//! altering the implementation to not require that `T` be `Sized`, such that it becomes
+//! `impl<T: Error + ?Sized> Error for Box<T>`, however, this altered implementation causes overlap
+//! between the `From` trait's `impl<T> From<T> for T` blanket implementation and Box's `impl<'a,
+//! E: Error + 'a> From<E> for Box<dyn Error + 'a>`.
+//!
+//! For more context on why `Box<dyn Error>` doesn't implement the `Error` trait, see 
+//! https://stackoverflow.com/questions/65151237/why-doesnt-boxdyn-error-implement-error. 
+//!
+//! `DynError` circumvents this overlap by being paired with a corresponding `DynResult` type that
+//! implements its own set of `FromResidual` impls (these exist so that `DynResult` works the same
+//! way with the `?` operator as `Result`). However, as a result, `DynError`s can only be
+//! constructed with `?` from arbitrary error types when paired with `DynResult`. Using a
+//! `Result<T, DynError>` will require manual conversion of error types due to it missing the
+//! `From` impl that is present on `Box<dyn Error>`. This is a bit of an ergonomic hit; it
+//! introduces a bit of runtime overhead as converting from a `DynError` to a `DynResult` requires
+//! downcasting at runtime.
```

Comment:
```suggestion
//! `From` impl that is present on `Box<dyn Error>`..
```

I still don't think I'd call this an ergonomic hit. If anything in practice the `DynResult` version will probably require less characters to write and otherwise be the same, since you end up writing things like `fn foo() -> DynResult<T>` instead of `fn foo() -> Result<T, BoxError>`. Leaving the error type out of the return signature is less typing overall. The issue is more that it is potentially confusing, and may cause some less than ideal diagnostics when people accidentally use the wrong Result type.

Also, I don't think we should focus on the additional runtime cost at all. It's super negligible, because the downcast is just one extra if statement and it will only ever go down the true branch when going from `DynResult` -> `Result<T, BoxError>` -> `DynResult`. In this case the real runtime cost is the allocation that we have to throw away from the `DynError` that was moved into a `Box` that we then have to extract from the `Box` to copy it back to the stack, and even then its still only copying two u64's worth of data.

My worry is that the comment as it's currently written is going to imply that the cost is much worse than it actually is and scare people away from using this as if it is less efficient than `Box<dyn Error>`, when in reality it is basically the same, since the heap allocation completely dwarfs the cost of the downcast.